core/primitives-core/src/serialize.rs

```diff
@@ -158,109 +158,136 @@ fn test_base_bytes_format() {
     assert_de_error::<Test>("{\"field\":null}");
 }
 
-pub mod u64_dec_format {
+/// Serialises number as a string; deserialises either as a string or number.
+///
+/// This format works for `u64`, `u128`, `Option<u64>` and `Option<u128>` types.
+/// When serialising, numbers are serialised as decimal strings.  When
+/// deserialising, strings are parsed as decimal numbers while numbers are
+/// interpreted as is.
+pub mod dec_format {
     use serde::de;
-    use serde::{Deserialize, Deserializer, Serializer};
+    use serde::{Deserializer, Serializer};
 
-    pub fn serialize<S>(num: &u64, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(&format!("{}", num))
+    /// Abstraction between integers that we serialise.
+    pub trait DecType: Sized {
+        /// Formats number as a decimal string; passes `None` as is.
+        fn serialize(&self) -> Option<String>;
+
+        /// Constructs Self from a `null` value.  Returns error if this type
+        /// does not accept `null` values.
+        fn try_from_unit() -> Result<Self, ()> {
+            Err(())
+        }
+
+        /// Tries to parse decimal string as an integer.
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError>;
+
+        /// Constructs Self from a 64-bit unsigned integer.
+        fn from_u64(value: u64) -> Self;
     }
 
-    pub fn deserialize<'de, D>(deserializer: D) -> Result<u64, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let s = String::deserialize(deserializer)?;
-        u64::from_str_radix(&s, 10).map_err(de::Error::custom)
+    impl DecType for u64 {
+        fn serialize(&self) -> Option<String> {
+            Some(self.to_string())
+        }
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError> {
+            Self::from_str_radix(value, 10)
+        }
+        fn from_u64(value: u64) -> Self {
+            value
+        }
     }
-}
 
-#[test]
-fn test_u64_dec_format() {
-    #[derive(PartialEq, Debug, serde::Deserialize, serde::Serialize)]
-    struct Test {
-        #[serde(with = "u64_dec_format")]
-        field: u64,
+    impl DecType for u128 {
+        fn serialize(&self) -> Option<String> {
+            Some(self.to_string())
+        }
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError> {
+            Self::from_str_radix(value, 10)
+        }
+        fn from_u64(value: u64) -> Self {
+            value as u128
```

Comment:
```suggestion
            value.into()
```

makes it mildly easier to audit for conversions with non-trivial implications.